docs/source/concepts/time.rst

```diff
@@ -0,0 +1,90 @@
+   where ``lt_C`` is the ledger time of the transaction that created ``C``.
+
+#. **Skew**: ``rt_TX - skew_min <= lt_TX <= rt_TX + skew_max``,
+   where ``skew_min`` and ``skew_max`` are parameters defined by the ledger.
+
+Apart from that, no other guarantees are given on the ledger time.
+Time has therefore to be considered slightly fuzzy in DAML, with the fuzziness depending on the skew parameters.
+DAML applications should not interpret the value returned by :ref:`getTime <daml-ref-gettime>` as a precise timestamp.
+
+
+Record time
+***********
+
+The *record time* ``rt_TX`` is another property of a transaction.
+It is timestamp with microsecond resolution,
+and is assigned by the ledger when the transaction is recorded on the ledger.
+
+The DAML ledger model does not prescribe how the record time is assigned.
+
+Ledger time model
+*****************
+
+The *ledger time model* is the set of parameters used in the assignment and validation of ledger time.
+It consists of the following:
+
+#. ``skew_min`` and ``skew_max``, the bounds on the difference between ``lt_TX`` and ``rt_TX``.
+
+#. ``transaction_latency``, the average duration from the time a transaction is submitted from a participant to the ledger
+   until the transaction is recorded.
+
+
+The ledger time model is part of the ledger configuration and can be changed by ledger operators through the
+:ref:`SetTimeModel <com.daml.ledger.api.v1.admin.SetTimeModelRequest>` config management API.
+
+Assigning ledger time
+*********************
+
+The ledger time is assigned automatically by the participant.
+In most cases, DAML applications will not need to worry about ledger time and record time at all.
+
+For reference, this section describes the details of how the ledger time is assigned. 
+
+#. When submitting commands over the ledger API,
+   users can optionally specify a ``min_ledger_time_rel`` or ``min_ledger_time_abs`` argument.
+   This defines a lower bound for the ledger time in relative and absolute terms, respectively.
+
+#. The ledger time is set to the highest of the following values:
+
+   #. ``max(lt_C_1, ..., lt_C_n)``, the maximum ledger time of all contracts used by the given transaction
+   #. ``t_p``, the local time on the participant
+   #. ``t_p + min_ledger_time_rel``, if ``min_ledger_time_rel`` is given
+   #. ``min_ledger_time_abs``, if ``min_ledger_time_abs`` is given
+
+#. If the set of contracts used by the transaction depends on time,
+   then the command interpretation may need to be repeated until a suitable ledger time is found.
+
+#. At this point, the ledger time may lie in the future (e.g., if a large value for ``min_ledger_time_rel`` was given).
+   The participant waits until ``lt_TX - transaction_latency`` before it submits the transaction to the ledger - 
+   the intention is that the transaction is record at ``lt_TX == rt_TX``.
+
+Use the parameters ``min_ledger_time_rel`` and ``min_ledger_time_abs`` if you expect that
```

Comment:
My understanding is that this only helps you very little. You can use this to ensure that the transaction arrives roughly at rt_TX at the ledger but If the ledger also interprets the transaction (to validate it) you’ve lost. You need to increase the skew as well to handle commands that take a long time to interpret.

Reply:
That's a very good point; for such a ledger, submitting the transaction with roughly `lt = 2 * interpretation_time` should work, but the Ledger API server might get in your way, as its guess of the `transaction_latency` will be far off. I propose we make `transaction_latency` also settable by the user; its purpose is only informative anyways, and if the user has better information, we should let them provide it. Yes, it requires the user to understand the guts of the architecture, but unfortunately I think that they have to, if they want to understand when their transactions will get accepted by the ledger.

A related question is on what kind of stability guarantees do we want to give here. Are we committing to the processing pipeline described here? I don't think we should, especially since if the ledger implementation changes, I believe the application will unfortunately have to adapt slightly. This is worth a sentence or two too IMO.

Reply:
Good point, `min_ledger_time_*` can only be used to work around a very long interpretation time, but it won't save you if the validation time is also too long. 

There are currently no plans to change the ledger time model. In the future, we might improve it to better handle such large transactions, e.g., giving the user more control, or by better estimating the actual time it will take for the transaction to be committed (from data like the current network latency, current ledger load, or transaction size).

> Are we committing to the processing pipeline described here

No, both the ledger time model and the algorithm for assigning may be changed in the future. I have added a sentence.

As a side note, `min_ledger_time_*` gives you the ability to delay submissions, which is useful for internal tests dealing with contention.

Reply:
We did have users run into this issue I believe so I would at least mention that this might not be sufficient and you need to bump the skew as well.